dbms/src/DataStreams/ParquetBlockInputStream.cpp

```diff
@@ -0,0 +1,169 @@
+#include <algorithm>
+#include <iterator>
+#include <vector>
+
+// TODO: clear includes
+#include <Core/ColumnWithTypeAndName.h>
+#include <DataStreams/ParquetBlockInputStream.h>
+#include <DataTypes/DataTypeFactory.h>
+#include <IO/BufferBase.h>
+#include <IO/ReadBufferFromMemory.h>
+#include <IO/WriteHelpers.h>
+#include <ext/range.h>
+
+#include <arrow/buffer.h>
+#include <arrow/api.h>
+#include <arrow/io/api.h>
+#include <parquet/arrow/reader.h>
+#include <parquet/arrow/writer.h>
+#include <parquet/exception.h>
+
+
+namespace DB
+{
+
+ParquetBlockInputStream::ParquetBlockInputStream(ReadBuffer & istr_, const Block & header_)
+    : istr(istr_)
+    , header(header_)
+{
+}
+
+Block ParquetBlockInputStream::getHeader() const
+{
+    return header;
+}
+
+
+void ParquetBlockInputStream::readData(const IDataType & type, IColumn & column, ReadBuffer & istr, size_t rows/*, double avg_value_size_hint*/)
+{
+    IDataType::InputStreamGetter input_stream_getter = [&] (const IDataType::SubstreamPath &) { return &istr; };
+    // TODO: will a zero avg_value_size_hint work for all types?
+    type.deserializeBinaryBulkWithMultipleStreams(column, input_stream_getter, rows, /* avg_value_size_hint = */0, false, {});  // TODO: avg_value_size_hint
+
+    if (column.size() != rows)
+        throw Exception("Cannot read all data in ParquetBlockInputStream.", ErrorCodes::CANNOT_READ_ALL_DATA);
+}
+
+using NameToColumnPtr = std::unordered_map<std::string, std::shared_ptr<arrow::Column>>;
+
+Block ParquetBlockInputStream::readImpl()
+{
+    Block res;
+
+    if (istr.eof())
+        return res;
+
+    // TODO: rewrite this garbage
```

Comment:
Use `void copyData(ReadBuffer & from, WriteBuffer & to);` where to is `WriteBufferFromString`